app/utils/deep_link/index.ts

```diff
@@ -114,80 +123,89 @@ export async function handleDeepLink(deepLinkUrl: string, intlShape?: IntlShape,
-        }
+            if (path === 'messages') {
+                if (identifier.startsWith('@')) {
+                    return {type: DeepLink.DirectMessage, url: deepLinkUrl, data: {serverUrl, teamName, userName: identifier.substring(1)}};
+                }
 
-        match = new RegExp('(.*)\\/([^\\/]+)\\/messages\\/@(\\S+)').exec(url);
-        if (match) {
-            return {type: DeepLink.DirectMessage, url: deepLinkUrl, data: {serverUrl: match[1], teamName: match[2], userName: match[3]}};
+                return {type: DeepLink.GroupMessage, url: deepLinkUrl, data: {serverUrl, teamName, channelName: identifier}};
+            }
         }
 
-        match = new RegExp('(.*)\\/([^\\/]+)\\/messages\\/(\\S+)').exec(url);
-        if (match) {
-            return {type: DeepLink.GroupMessage, url: deepLinkUrl, data: {serverUrl: match[1], teamName: match[2], channelId: match[3]}};
+        const permalinkMatch = matchPermalinkDeeplink(url);
+        if (permalinkMatch) {
+            const {params: {serverUrl, teamName, postId}} = permalinkMatch;
+            return {type: DeepLink.Permalink, url: deepLinkUrl, data: {serverUrl, teamName, postId}};
         }
 
-        match = new RegExp('(.*)\\/plugins\\/([^\\/]+)\\/(\\S+)').exec(url);
-        if (match) {
-            return {type: DeepLink.Plugin, url: deepLinkUrl, data: {serverUrl: match[1], id: match[2], teamName: ''}};
+        const pluginMatch = match<{serverUrl: string; id: string}>(`:serverUrl(.*)/plugins/:id(${PLUGIN_ID_PATH_PATTERN})`)(url);
+        if (pluginMatch) {
+            const {params: {serverUrl, id}} = pluginMatch;
+            return {type: DeepLink.Plugin, url: deepLinkUrl, data: {serverUrl, teamName: '', id}};
         }
-    } catch {
+    } catch (err) {
         // do nothing just return invalid deeplink
     }
 
     return {type: DeepLink.Invalid, url: deepLinkUrl};
 }
 
-export function matchDeepLink(url?: string, serverURL?: string, siteURL?: string) {
+export function matchDeepLink(url: string, serverURL?: string, siteURL?: string) {
     if (!url || (!serverURL && !siteURL)) {
-        return '';
+        return null;
     }
 
     let urlToMatch = url;
     const urlBase = serverURL || siteURL || '';
     const parsedUrl = urlParse(url);
-    const parsedBase = urlParse(urlBase);
 
     if (!parsedUrl.protocol) {
         // If url doesn't contain site or server URL, tack it on.
         // e.g. <jump to convo> URLs from autolink plugin.
-        const match = new RegExp(escapeRegex(urlBase)).exec(url);
-        if (!match) {
+        const deepLinkMatch = new RegExp(escapeRegex(urlBase)).exec(url);
+        if (!deepLinkMatch) {
             urlToMatch = urlBase + url;
         }
     }
 
-    const finalUrl = urlParse(urlToMatch);
-    const baseSubpath = parsedBase.pathname.replace('/', '');
-    const baseHostname = parsedBase.hostname;
-    const urlSubpath = finalUrl.pathname.split('/')[1];
```

Comment:
There was a bug here that couldn't handle deep subpaths. e.g. `example.com/subpath/deepersubpath/etc/teamName/pl/…`